buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesGraphPlugin.java

```diff
@@ -41,5 +38,13 @@ public void apply(Project project) {
                 t.setUrl(url);
             });
         });
+
+        project.getGradle().getTaskGraph().whenReady(graph -> {
+            if (graph.getAllTasks().stream().anyMatch(t -> t instanceof DependenciesGraphTask)) {
+                if (url == null || token == null) {
+                    throw new GradleException("The environment variables SCA_URL and SCA_TOKEN need to be set before this task is run");
```

Comment:
Let's make the message `"The environment variables SCA_URL and SCA_TOKEN need to be set before task " + t.getPath() +" can run"` as it won't be clear to the user what we mean by "this task" w/o that context.

Reply:
My thinking was that one either 

- runs the task for the whole project (`gradlew -q dependenciesTask`) so that in the DependenciesGraphPlugin ,if we filter for `instanceof DependenciesGraphTask` we will get all the subprojects. Then which one do you add to the message ( which will be shown N times ) 
- runs the task for one project (`gradlew -q x-pack:plugin:core:dependenciesTask` ) so they know what "this task" is

Reply:
I would just error on the first task it finds. In any case the "solution" is the same for all the others so it's not like we are short-circuiting some other error that'll encounter.

Reply:
Sure, makes sense. It's just a comment to whether printing the task name in the message is helpful here or not. So it's either someone ran this for all subprojects and we will show a random `X:dependenciesTask` in the error message, or someone ran this as a specific task, so they know what "this task" is (context is there), so showing the task is not that helpful

I can add this if you think we should though, happy to wrap this up either way we decide

Reply:
Yeah, you could argue in this case folks will be explicitly running those tasks, but I think it's worthwhile to confirm, since there are likely other upstream tasks run here and when in doubt, better to be explicit :smile:

Reply:
I still think that printing 

```
* What went wrong:
Failed to notify task execution graph listener.
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :x-pack:plugin:core:dependenciesGraph can run
....
....
....
....
N times
```

when someone just ran `./gradlew -q x-pack:plugin:core:dependenciesGraph` and

```
* What went wrong:
Failed to notify task execution graph listener.
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :benchmarks:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :benchmarks:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :benchmarks:dependenciesGraph can run
> The environment variables SCA_URL and SCA_TOKEN need to be set before task :benchmarks:dependenciesGraph can run
...
...
...
N times
```

when someone ran `./gradlew -q dependenciesGraph` , 

is a bit on the extreme explicitness side, but I'll comply and pick another hill :)

Reply:
Haha, you are absolutely right. The reason for this is because we are registering that task graph listener multiple times, in fact, every time the plugin is applied. The the task execution graph is "global" so you actually only need do this once. We can solve this by putting that logic into a separate plugin, and applying that plugin only to the root project. Plugin application is idempotent so we can apply it to the root project over and over but it'll only actually evaluate once. We do this elsewhere when we want to register "global" listeners by a) defining a plugin in an inner class b) adding some logic to that plugin to error if we mistakenly apply it to a project that isn't the root project c) apply that new plugin from the "outer" plugin

https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java#L138

Reply:
Nice, thanks! I took a swing at this ( Apologies this is dragging but it's low priority - since we got the first graph in manually - and I'm taking long cycles before spending time on it again )